### ISSUE TITLE
[SDK] How to override with Dagger injections, when everything is INTERNAL

### ISSUE THREAD

**Issue by tladesignz** [vector-im/element-android#4724]
### Description

Keanu would like to inject custom HTTP headers in requests (specifically during account registration).

Of course we could do that with Retrofit interceptors, but how to do that, when everything is internal?

https://gitlab.com/keanuapp/keanuapp-ios/-/issues/236#note_645003296

As far as I understand, I would need to do my own `DefaultAuthenticationService` and inject that in my own `Matrix` class. But all the other stuff which I don't want to touch is in the internal namespace, and therefore inaccessible. I would need to copy all the concrete implementations of the interfaces, too.

Or do I miss something here?

**Comment by ouchadam** [comment]
you're correct, the internal visibility limits almost of all of the extensibility of the SDK (I assume by design)
as far as I'm aware the only ways currently to modify the SDK are to import the code locally or submit a change to this repository 

you might be able to expose the internal kotlin classes with Java wrappers but it would be very hacky and you'd still end up copying lots of classes

**Comment by bmarty** [comment]
The SDK could expose an API, or something in [MatrixConfiguration](https://github.com/vector-im/element-android/blob/main/matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt#L23), to let the clients easily add any static headers they want to all the requests. Would that fit you needs @tladesignz ?

**Comment by tladesignz** [comment]
Folks, thank you very much for the quick replies!

> The SDK could expose an API, or something in [MatrixConfiguration](https://github.com/vector-im/element-android/blob/main/matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt#L23), to let the clients easily add any static headers they want to all the requests. Would that fit you needs @tladesignz ?

Very much so, yes! :-)
